Build protos via docker

ajm188 opened the pull request:
This PR creates a dockerfile and build script for generating the go code from the .proto files.

I was playing around with the protos locally when I discovered that `make proto` did not work with the way I had cloned vitess (I think) (errors like `/bin/bash: line 0: cd: /../../../src: No such file or directory`).

After poking at it for a bit, I ended up deciding (but let me know if you think differently!) that going with a dockerized solution would probably be simpler to maintain. The main benefits I see are that we pin the version of `protoc` (we could also pin `protoc-gen-go` and `goimports` if we wanted) and that it doesn't rely on VTTOP or VTROOT in any way; you just need a vitess checkout and docker installed.

Testing:

```
❯ make proto
Mon Aug 10 13:41:59 EDT 2020: Compiling proto definitions
Sending build context to Docker daemon  210.3MB
Step 1/8 : FROM golang:1.14-buster
 ---> baaca3151cdb
Step 2/8 : RUN apt-get update &&     DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends         build-essential         ca-certificates         curl         unzip         wget
 ---> Using cache
 ---> b5fc7362d68c
Step 3/8 : ARG PROTOC_GEN_GO_VERSION=v1.3.1
 ---> Using cache
 ---> 4c3b0629bd2f
Step 4/8 : RUN mkdir proto-builder &&     cd proto-builder &&     go mod init proto-builder &&     GO111MODULE=on go get         golang.org/x/tools/cmd/goimports         github.com/golang/protobuf/protoc-gen-go@${PROTOC_GEN_GO_VERSION}
 ---> Using cache
 ---> 1eba290a1f52
Step 5/8 : ARG PROTOC_VERSION=3.12.4
 ---> Using cache
 ---> 21ade038f4b4
Step 6/8 : RUN curl -LO https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOC_VERSION}/protoc-${PROTOC_VERSION}-linux-x86_64.zip &&     unzip protoc-${PROTOC_VERSION}-linux-x86_64.zip -d /protoc &&     mv /protoc/bin/protoc /usr/local/bin/ &&     rm -rf protoc-${PROTOC_VERSION}-linux-x86_64.zip
 ---> Using cache
 ---> c30391a299f6
Step 7/8 : ADD proto proto
 ---> Using cache
 ---> 1a616f2d15ad
Step 8/8 : CMD [ "./proto/build.sh" ]
 ---> Using cache
 ---> 0fe79cf99c36
Successfully built 0fe79cf99c36
Successfully tagged proto-builder:latest
Building automation.proto ...
Building automationservice.proto ...
Building binlogdata.proto ...
Building binlogservice.proto ...
Building logutil.proto ...
Building mysqlctl.proto ...
Building query.proto ...
Building queryservice.proto ...
Building replicationdata.proto ...
Building tableacl.proto ...
Building tabletmanagerdata.proto ...
Building tabletmanagerservice.proto ...
Building throttlerdata.proto ...
Building throttlerservice.proto ...
Building topodata.proto ...
Building vschema.proto ...
Building vtctldata.proto ...
Building vtctlservice.proto ...
Building vtgate.proto ...
Building vtgateservice.proto ...
Building vtrpc.proto ...
Building vttest.proto ...
Building vttime.proto ...
Building vtworkerdata.proto ...
Building vtworkerservice.proto ...
Building workflow.proto ...
```

~~I'm not sure what version of protoc was used to generate these last time, because I definitely see changes, [vtgate.pb.go, for example](https://gist.github.com/ajm188/a92a3f406401d04129f09e08fc9c36cd). I would guess pinning the protoc version lower could let us merge this change _without_ impacting the protos (and then later we can bump the version and update the protos :D)~~

I pinned proto-gen-go to 1.3.2 (found [here](https://github.com/vitessio/vitess/blob/master/go.mod#L28)), and now the only diff is the following in go/vt/proto/binlogdata/binlogdata.pb.go:

```diff
diff --git a/go/vt/proto/binlogdata/binlogdata.pb.go b/go/vt/proto/binlogdata/binlogdata.pb.go
index 5df50ab9d..75a2108d4 100644
--- a/go/vt/proto/binlogdata/binlogdata.pb.go
+++ b/go/vt/proto/binlogdata/binlogdata.pb.go
@@ -1383,9 +1383,6 @@ func (m *VEvent) GetJournal() *Journal {
 
 func (m *VEvent) GetDml() string {
        if m != nil {
-               if m.Statement != "" {
-                       return m.Statement
-               }
                return m.Dml
        }
        return ""
```

dkhenry commented on Makefile:
```diff
@@ -149,20 +149,14 @@ PROTO_SRC_NAMES = $(basename $(notdir $(PROTO_SRCS)))
 # This rule rebuilds all the go files from the proto definitions for gRPC.
-proto: $(PROTO_GO_OUTS)
-
+proto: $(PROTO_SRCS)
```
I don't think this should take over the proto target. At a minimum we should keep the old target to build the proto's on the local system if the developer would like to

ajm188 replied:
Are you able to build those protos? The main reason I started making this change is because I could never get `make proto` to work for me.

dkhenry commented on Makefile:
```diff
@@ -149,20 +149,14 @@ PROTO_SRC_NAMES = $(basename $(notdir $(PROTO_SRCS)))
-		goimports -w $(VTROOT)/go/vt/proto/$${name}/$${name}.pb.go; \
-	done
+	docker build -t proto-builder -f docker/proto/Dockerfile .
+	docker run --rm --mount type=bind,src=$(shell pwd)/go/vt/proto/,dst=/go/vt/proto/ proto-builder
```
This is likey why you have to sudo below. Instead of bind mounting just use `-v` to mount the directory and run the container as the current user.

shlomi-noach replied:
this `mount` seems to have the unfortunate effect of making chmod-ing directories to `777`

ajm188 replied:
Yeah, I'm going to push up a commit switching to the `-v` style

dkhenry commented on docker/proto/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+RUN curl -LO https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOC_VERSION}/protoc-${PROTOC_VERSION}-linux-x86_64.zip && \
+    unzip protoc-${PROTOC_VERSION}-linux-x86_64.zip -d /protoc && \
+    mv /protoc/bin/protoc /usr/local/bin/ && \
+    rm -rf protoc-${PROTOC_VERSION}-linux-x86_64.zip
```
I am not sure what the end use of this container would be, but I would support having it be built once and pushed to the vitess repo on docker hub. That would prevent the fetching of these dependencies every time we rebuild proto's

ajm188 replied:
Pushing this up somewhere sounds like a great idea! Do you think it should go under vitess's dockerhub, or somewhere else?

shlomi-noach commented:
Definitely works for me, and more reliable and consistent than my `protoc` execution.

sougou commented:
Can you check if #6580 solves the problem?

shlomi-noach commented:
on my machine, #6580 does not work as expected:
```shell
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
--go_out: protoc-gen-go: plugins are not supported; use 'protoc --go-grpc_out=...' to generate gRPC
...
(infinite)
```

ajm188 commented:
I believe I've fixed all the permissions issues with the resulting. generated protos, and also moved the docker-based proto generation to a separate make target, which I believe were the two biggest comments on this change.

The main thing I think is outstanding is [this comment](https://github.com/vitessio/vitess/pull/6548#discussion_r468960146) about pushing the builder image up to a registry.

deepthi commented:
@ajm188 for some reason GH won't let me re-run the failed tests. can you try?

ajm188 commented:
Hmmm, it looks like I'm missing the "Re-run all jobs" button, too. Did something change with permissions?

deepthi commented:
> Hmmm, it looks like I'm missing the "Re-run all jobs" button, too. Did something change with permissions?

Nope. I can do that on other PRs just not on this one. I'm going to try close followed by re-open. If that doesn't work maybe you could close this PR and open another one from the same branch.

ajm188 commented:
I see the button again after you did that.

deepthi commented:
a few things still to be clarified/resolved:
- How would this fit into the dev process for someone who is relying on a local build?
- We have to validate that the Makefile changes don't break `make proto` for environments where it currently works
- Decide on a name/tag for the new docker image (vitess/proto?) and how it gets updated when the source files change

ajm188 commented:
> How would this fit into the dev process for someone who is relying on a local build?

Right now, anyone that has a local environment setup that can build the protos should be unaffected. For people that are having issues building protos (@shlomi-noach's issue, for example, is fixed by downgrading his install of protoc-gen-go to v1.23.0) can, assuming they have docker working locally, just switch to `make docker-proto` rather than change whatever is off in their environment (i.e. if Shlomi needs the newer version of protoc-gen-go for some other project).

(I think eventually we should build everything in docker because it feels simpler to me for newcomers with nothing set up, but that doesn't have to be the case with this change).

> We have to validate that the Makefile changes don't break make proto for environments where it currently works

On my branch, I did the following:

```
❯ find go/vt/proto -iname '*.pb.go' -exec rm -rf {} \;
❯ make proto
❯ git d | cat
diff --git a/go/vt/proto/binlogdata/binlogdata.pb.go b/go/vt/proto/binlogdata/binlogdata.pb.go
index 5df50ab9d..75a2108d4 100644
--- a/go/vt/proto/binlogdata/binlogdata.pb.go
+++ b/go/vt/proto/binlogdata/binlogdata.pb.go
@@ -1383,9 +1383,6 @@ func (m *VEvent) GetJournal() *Journal {
 
 func (m *VEvent) GetDml() string {
        if m != nil {
-               if m.Statement != "" {
-                       return m.Statement
-               }
                return m.Dml
        }
        return ""
```

That extra diff seems to be added manually in https://github.com/vitessio/vitess/commit/250caba0fabaefa952cb05b615015bf32ee0735f, so the non-docker version seems to be working fine.

> Decide on a name/tag for the new docker image (vitess/proto?) and how it gets updated when the source files change

`vitess/proto` sounds good to me! As for how it gets updated, I am thinking anytime someone opens a PR that touches something in `proto/`, they, or the person merging should build+push the image (maybe I should add something in ReleaseInstructions.md for this? Or maybe a separate doc). Once we have the docker images building and pushed (I may need some help, or someone to do it for me to push to vitess/*) we can update the Makefile to use that instead of always building locally.

shlomi-noach commented:
I'm still unable to `make proto` and am happy to merge this PR (which adds a `make docker-proto`). Are there any objections?

shlomi-noach commented:
I've cleared @dkhenry 's request for change since that change was applied.

shlomi-noach commented:
Actually, I'm having trouble with the docker build, too. I've just opened #7041 which seems to work well (for me).

ajm188 commented:
What issue did you run into?

shlomi-noach commented:
> What issue did you run into?

Sorry, I should have been more detailed. And... now that I came back to this, I can't reproduce and everything seems fine. What I got yesterday are successful builds but strange outputs; that is `*.pb.go` did not match well my `*.proto` changes. Perhaps I had some git/merge confusion yesterday.

shlomi-noach commented:
All, `make proto` is still broken. I'm using this PR as a place to revive the discussion. Which way do we want to go?

ajm188 commented:
@shlomi-noach how is `make proto` broken for you? I've been running `source ./dev.env && make proto` since #7519 landed and have had no issues.

ajm188 commented:
There was an issue I was having with the new generator not being findable and I asked @vmg in vitess slack [here](https://vitess.slack.com/archives/CMDJ2KFEZ/p1614356560012800)

shlomi-noach commented:
Now resolved... thanks to @deepthi.

I had this going on:
```
$ make proto
Sun 07 Mar 2021 17:26:52 IST: Installing minimal dependencies
skipping protoc install. remove /home/shlomi/dev/github/planetscale/vitess-dopple/dist/vt-protoc-3.6.1 to force re-install.
/usr/bin/etcd
etcd already installed
/home/shlomi/dev/github/planetscale/vitess-dopple/bin/k3s

bootstrap finished - run 'make build' to compile
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
protoc-gen-gofast: program not found or is not executable
--gofast_out: protoc-gen-gofast: Plugin failed with status code 1.
...
```

and it seems I need to manually run `go install github.com/gogo/protobuf/protoc-gen-gofast`; it was supposed to run as dependency for `make proto`, but after running this manually everything seems fine.
